### PR TITLE
Documenting Migrators and Migrations.

### LINKED ISSUE #1355: Document migrators
In "Knowledge Base" and elsewhere, we talk about migrations (grep "migration" or "migrator" in the raw doc files, ex: under CUDA/Apple Silicon sections), but AFAIK we never carefully explain to maintainers 
- What is a migrator/migration and what does it do?
- What issues the "Rebuild for ..." PRs? (Ans: the migrator)
- How to write one, and where does it go? (Ans: see [here](https://conda-forge.org/docs/maintainer/pinning_deps.html#updating-package-pins))
- When can (and why would) I reject a migration PR?
- How to resolve rerendering conflicts between the migrators and a local `conda_build_config.yaml`?

The last item is the most painful and subtle pitfall. The applied order is global pinning -> local pinning from `conda_build_config.yaml` -> migrators (see https://gitter.im/conda-forge/conda-forge.github.io?at=60335ce247585464db940db3), but maybe @isuruf or @beckermr or someone from @conda-forge/core can confirm.
Often if there is a conflict, the rerendering would fail obscurely and it's hard to debug.

@viniciusdc This may require more prerequisite knowledge of the Conda-Forge infrastructure than other documentation tasks, so I am not sure if it's suitable for the outreachy participants and I'll leave it for you and the core team to decide 🙂

### PR DESCRIPTION
**Pull request by ssurbhi560**
PR Checklist:
- [x] note any issues closed by this PR with [closing keywords](https://help.github.com/articles/closing-issues-using-keywords)
- [x] put any other relevant information below
Closes #1355, #862 and #737

### CODE REVIEW

**Comment by beckermr** [comment]
I think we need the changes from main for this to pass.

File: src/maintainer/knowledge_base.rst
Comment (jaimergp) [inline src/maintainer/knowledge_base.rst:1525]: Not all packages that depend on a pinned package need to be rebuilt. We only need to rebuilt those that impose an ABI dependency (usually via linking / dynamic libraries, but sometimes also header libraries like pybind11). I am not sure of what the right terminology is here, though, but the rebuild only concerns packages which depend on pinned packages in their `host` section.

```suggestion
When any changes are made in the global pinnings of a package, then the entire stack of the packages that need that package on their `host` section would need to be updated and rebuilt.
```

File: src/maintainer/knowledge_base.rst
Comment (jaimergp) [inline src/maintainer/knowledge_base.rst:1526]:
```suggestion
Doing it manually could be quite tedious, and that's where migrations come to help. Migrations automate the process of submitting changes to a feedstock and are an integral part of ``regro-cf-autotick-bot``'s duties.
```

File: src/maintainer/knowledge_base.rst
Comment (jaimergp) [inline src/maintainer/knowledge_base.rst:1528]: There are several types of migrations. Most of the time we only need an updated pin, which can be triggered via migration files. However there are others. I'd mention the `cf-scripts` docs: https://regro.github.io/cf-scripts/migrators.html too

File: src/maintainer/knowledge_base.rst
Comment (jaimergp) [inline src/maintainer/knowledge_base.rst:1544]: I would use stronger wording here. We kindly ask people do NOT close PRs if they are unwilling to work on it. Reasoning:

- The bot will assume another PR implementing the migration was merged instead, letting the migration continue iterating on the graph; however downstream dependents will fail because the parent (the one we closed the PR of) didn't really get built.
- Somebody might pass by and help if they see an open PR

File: src/maintainer/knowledge_base.rst
Comment (kathatherine) [reply src/maintainer/knowledge_base.rst:1526]: I'd also suggest using present tense where possible.

```suggestion
Doing it manually can be quite tedious, and that's where migrations come to help. Migrations are used to change how different feedstocks/packages operate and are an integral part of the ``regro-cf-autotick-bot``'s duties.
```

File: src/maintainer/knowledge_base.rst
Comment (kathatherine) [inline src/maintainer/knowledge_base.rst:1532]: Same here with present tense.

```suggestion
To propose a migration in one or more pins, a PR is issued into the pinning feedstock using a yaml file expressing the changes to the global pinning file in the migrations folder. 
Once the PR is squashed and merged, the bot takes the yaml file to calculate which packages need to be rebuilt and starts issuing PRs with the new pinning as a local pinning file to the feedstocks.

Usually, the bot generates these migrations automatically. However, when a pin is first made or added, one may need to be added by hand. To do this, follow the steps mentioned in `Updating package pins <https://conda-forge.org/docs/maintainer/pinning_deps.html#updating-package-pins>`__.
```

File: src/maintainer/knowledge_base.rst
Comment (kathatherine) [inline src/maintainer/knowledge_base.rst:1540]: Does the "it" in that second bullet point refer to the node or something else?

File: src/maintainer/knowledge_base.rst
Comment (kathatherine) [inline src/maintainer/knowledge_base.rst:1544]: If you keep any of the language in this sentence after going through Jaime's comment above, here's some suggestions for the grammar.

```suggestion
If you close the PR, it makes the bot consider that PR as merged and that can cause inconsistencies with downstream feedstocks.
```

**Review by kathatherine — CHANGES_REQUESTED** [review]
Thanks, Surbhi! Here's some edits.

File: src/maintainer/knowledge_base.rst
Comment (ssurbhi560) [reply src/maintainer/knowledge_base.rst:1540]: @kathatherine, it refers to node. Should I replace "It" with "the node"?

Comment (kathatherine) [reply src/maintainer/knowledge_base.rst]: @ssurbhi560 Yes, please! Would help with clarity.

File: src/maintainer/knowledge_base.rst
Comment (kathatherine) [inline src/maintainer/knowledge_base.rst:1530]:
```suggestion
Once the PR is squashed and merged, the migrator bot uses the yaml file to calculate which packages need to be rebuilt and starts issuing PRs with the new pinning as a local pinning file to the feedstocks.
```

File: src/maintainer/knowledge_base.rst
Comment (kathatherine) [inline src/maintainer/knowledge_base.rst:1546]:
```suggestion
If you close the PR, it makes the bot think that another PR implementing the migration is merged instead, letting the migration continue iterating on the graph; however, the downstream dependents fail because the parent (the one we closed the PR of) didn’t really get built.
```

**Review by kathatherine — CHANGES_REQUESTED** [review]
Found some more edits to make things more clear. Thanks!

**Review by kathatherine — APPROVED** [review]
Looks good to me! Thanks for letting me iterate over the words with you.

**Review by beckermr — APPROVED** [review]

File: src/maintainer/knowledge_base.rst
Comment (leofang) [inline src/maintainer/knowledge_base.rst:1541]: Why do we have to mention "squash" here? Please don't... It's a very bad practice against tracking changes, sometimes a migration PR needs additional work by the maintainers. I understand there are "purists" out there who would defend the cleanness of the commit history, but really it's unnecessary to advocate.
```suggestion
Once the PR is merged, the migrator bot uses the yaml file to calculate which packages need to be rebuilt and starts issuing PRs with the new pinning as a local pinning file to the feedstocks.
```

Comment (leofang) [reply src/maintainer/knowledge_base.rst:1541]: (This sentence is also a bit weird to me, IIRC we build the whole dependency graph first, and then the bot walks through the graph to migrate all nodes one by one. @beckermr can chime in here.)

File: src/maintainer/knowledge_base.rst
Comment (leofang) [inline src/maintainer/knowledge_base.rst:1547]: I noticed we use both lowercase "yaml" and uppercase "YAML" in this section. Can we unify them?

File: src/maintainer/knowledge_base.rst
Comment (leofang) [inline src/maintainer/knowledge_base.rst:1549]: Up to this point we haven't mentioned anything about graphs or nodes, so we should be clear what we're referring to:
```suggestion
  3. A migration PR is issued for a node (a feedstock) only if:
```

File: src/maintainer/knowledge_base.rst
Comment (leofang) [inline src/maintainer/knowledge_base.rst:1554]:
```suggestion
  4. Process 3 continues until the migration is complete and the change is applied to the global pinning file via a final PR. After this step, we say this migration is closed out.
```

File: src/maintainer/knowledge_base.rst
Comment (leofang) [inline src/maintainer/knowledge_base.rst:1548]:
```suggestion
  2. Once the PR is merged, the bot picks it up, builds a migrator graph, and begins the migration process.
```

File: src/maintainer/knowledge_base.rst
Comment (leofang) [inline src/maintainer/knowledge_base.rst:1557]: Note: here we talk about the graph.

Comment (leofang) [reply src/maintainer/knowledge_base.rst:1557]: nit:
```suggestion
If you close the PR, it makes the bot think that another PR implementing the migration is merged instead, letting the migration continue iterating on the graph; however, the downstream dependents fail because the parent (the one we closed the PR of) didn’t really get rebuilt.
```

**Comment by jaimergp** [comment]
Let's address those comments by @leofang first, @kathatherine

After that I think we'll be good.

File: src/maintainer/knowledge_base.rst
Comment (ssurbhi560) [reply src/maintainer/knowledge_base.rst:1541]: Hi @leofang! Thank you so much for reviewing the PR. 
Rephrasing it to - "Once the PR is merged, the dependency graph is built. After that, the bot walks through the graph and migrates all the nodes (feedstocks) one by one and issues PRs for those feedstocks." would this be correct? 
@kathatherine, can you also please take a look and let me know if something needs to be changed in this?

Comment (kathatherine) [reply src/maintainer/knowledge_base.rst:1541]: @ssurbhi560 Let's avoid two "and"s in the same list.

"Once the PR is merged, the dependency graph is built. After that, the bot walks through the graph, migrates all the nodes (feedstocks) one by one, and issues PRs for those feedstocks."

Would also love @beckermr's check of this.

Comment (beckermr) [reply src/maintainer/knowledge_base.rst]: Yes @leofang is correct. We issue PRs in topological order in the graph and downstream packages have to wait for their upstream parents before they get a PR.

**Review by kathatherine — APPROVED** [review]
Thanks, Surbhi!

**Review by leofang — APPROVED** [review]
LGTM, thx @ssurbhi560!

**Comment by kathatherine** [comment]
@beckermr @jaimergp Should be good to merge!

**Pull request merged**
